File: ticket_office/cs/TicketOffice/Domain/Train.cs
namespace TicketOffice.Domain;

// Note: no ID - we are not in the same bounded context
// as train_data
public class Train
{
    private Dictionary<string, Seat> _seats;

    public Train()
    {
        _seats = new Dictionary<string, Seat>();
    }
    public List<Seat> Seats()
    {
        return _seats.Values.ToList();
    }

    public void Add(Seat seat)
    {
        _seats[seat.Id] = seat;
    }

    public void Book(string seatId, string bookingReference)
    {
        var seat = GetSeat(seatId);
        seat.Book(bookingReference);
    }

    public Seat GetSeat(string id)
    {
        var res = _seats.GetValueOrDefault(id);
        if (res == null)
        {
            throw new NoSuchSeatException(id);
        }
        return res;
    }
}

File: java/src/main/java/fr/arolla/trainreservation/domain/Train.java
package fr.arolla.trainreservation.domain;

import fr.arolla.trainreservation.infra.Reservation;

import java.util.HashMap;
import java.util.List;

public class Train {
  private final HashMap<String, Seat> seats = new HashMap<>();

  public Train(List<Seat> seats) {
    for (var seat : seats) {
      this.seats.put(seat.id(), seat);
    }
  }

  public List<Seat> getSeats() {
    return seats.values().stream().toList();
  }

  public void book(Reservation reservation) {
    String bookingReference = reservation.bookingReference();
    List<String> seats = reservation.seats();
    for (var seatId : seats) {
      bookSeat(seatId, bookingReference);
    }
  }

  public void bookSeat(String seatId, String bookingReference) {
    var seat = this.seats.get(seatId);
    if (seat == null) {
      throw new NoSuchSeatException(seatId);
    }
    seat.book(bookingReference);
  }
}
